react-components/src/components/Reveal3DResources/calculatePointCloudModelsStyling.tsx

```diff
@@ -0,0 +1,228 @@
+import { EMPTY_ARRAY } from '../../utilities/constants';
+import { type AnnotationId, type PointCloudAnnotationModel } from '../CacheProvider/types';
+import { type PointCloudAnnotationCache } from '../CacheProvider/PointCloudAnnotationCache';
+
+export type StyledPointCloudModelAddOptions = {
+  addOptions: PointCloudModelOptions;
+  styleGroups: AnnotationIdStylingGroup[];
+  defaultStyle: PointCloudAppearance;
+};
+
+type PointCloudAddOptionsWithStyleGroups = {
+  addOptions: PointCloudModelOptions;
+  styleGroups: AnnotationIdStylingGroup[];
+};
+
+export type AnnotationModelDataResult = {
+  model: PointCloudModelOptions;
+  annotationModel: PointCloudAnnotationModel[];
+};
+
+export const calculatePointCloudStyling = async (
+  models: PointCloudModelOptions[],
+  instanceGroups: AssetStylingGroup[],
+  defaultResourceStyling: DefaultResourceStyling | undefined,
+  pointCloudAnnotationCache: PointCloudAnnotationCache
+): Promise<StyledPointCloudModelAddOptions[]> => {
+  const styledPointCloudModels = await calculateMappedPointCloudStyling(
+    models,
+    defaultResourceStyling?.pointcloud?.mapped,
+    pointCloudAnnotationCache
+  );
+  const modelInstanceStyleGroups = await calculateInstanceStyling(
+    models,
+    instanceGroups,
+    pointCloudAnnotationCache
+  );
+
+  const combinedStyledPointCloudModels = groupStyleGroupByModel(
+    [...styledPointCloudModels, ...modelInstanceStyleGroups],
+    defaultResourceStyling?.pointcloud?.default
+  );
+  return combinedStyledPointCloudModels;
+};
+
+async function calculateInstanceStyling(
+  models: PointCloudModelOptions[],
+  instanceGroups: AssetStylingGroup[],
+  pointCloudAnnotationCache: PointCloudAnnotationCache
+): Promise<PointCloudAddOptionsWithStyleGroups[]> {
+  const pointCloudAnnotationMappings = await getPointCloudAnnotationMappingsForModels(
+    models,
+    pointCloudAnnotationCache
+  );
+
+  const styledModels = pointCloudAnnotationMappings?.map((annotationMappings) => {
+    return calculateAnnotationMappingModelStyling(instanceGroups, annotationMappings);
+  });
+
+  return styledModels;
+}
+
+function calculateAnnotationMappingModelStyling(
+  instanceGroups: AssetStylingGroup[],
+  annotationMapping: AnnotationModelDataResult
+): PointCloudAddOptionsWithStyleGroups {
+  const styleGroups = instanceGroups
+    .map((group) => {
+      return getMappedStyleGroupFromAssetIds(annotationMapping, group);
+    })
+    .filter((styleGroup): styleGroup is AnnotationIdStylingGroup => styleGroup !== undefined);
+
+  return { addOptions: annotationMapping.model, styleGroups };
+}
+
+function getMappedStyleGroupFromAssetIds(
+  annotationMapping: AnnotationModelDataResult,
+  instanceGroup: AssetStylingGroup
+): AnnotationIdStylingGroup | undefined {
+  const uniqueAnnotationIds = new Set<number>();
+  const assetIdsSet = new Set(instanceGroup.assetIds.map((id) => id));
+
+  const matchedAnnotationModels = annotationMapping.annotationModel.filter((annotation) => {
```

Comment:
can we use `filter` and `ìncludes`to simplfy filtering of `ànnotationModel`?

Reply:
Simplified a little bit, at least